### PR TITLE
Azure: Modify azure provisioner to use default credentials 

### PR DESCRIPTION
**Pull request by kartikjoshi21**
Use default credentials to authenticate in azure go sdk, so that
it can choose what to authenticate from.
The order of authentication will be:
1. Environment variables
2. Managed identity
3. Azure cli auth

Also modify aks config to allow system assigned service principal in case of azure cli auth.

### CODE REVIEW

File: test/provisioner/provision_azure.properties [hunk 1]
```diff
@@ -11,3 +12,4 @@ SSH_KEY_ID="id_rsa.pub"
 CLOUD_PROVIDER="azure"
 AZURE_IMAGE_ID=""
 SSH_USERNAME=""
+AZURE_CLI_AUTH="no"
```
Comment (surajssd) [inline test/provisioner/provision_azure.properties:15]: let's stick to `true` and `false`?

File: test/provisioner/provision_azure_initializer.go [hunk 2]
```diff
@@ -109,31 +118,41 @@ func initAzureProperties(properties map[string]string) error {
...
+		return err
+	}
+	virtualNetworksClient := networkClientFactory.NewVirtualNetworksClient()
+	subnetsClient := networkClientFactory.NewSubnetsClient()
 
-	subnetClient := network.NewSubnetsClient(AzureProps.SubscriptionID)
-	subnetClient.Authorizer = authorizer
+	computeClientFactory, err := armcompute.NewClientFactory(AzureProps.SubscriptionID, cred, nil)
+	if err != nil {
+		log.Fatal(err)
```
Comment (surajssd) [inline test/provisioner/provision_azure_initializer.go:141]: return err?

File: test/provisioner/provision_azure_initializer.go [hunk 3]
```diff
@@ -109,31 +118,41 @@ func initAzureProperties(properties map[string]string) error {
...
+	if err != nil {
+		return err
+	}
+	resourceGroupClient := resourcesClientFactory.NewResourceGroupsClient()
 
-	aksClient := containerservice.NewManagedClustersClient(AzureProps.SubscriptionID)
-	aksClient.Authorizer = authorizer
+	networkClientFactory, err := armnetwork.NewClientFactory(AzureProps.SubscriptionID, cred, nil)
+	if err != nil {
+		return err
```
Comment (surajssd) [inline test/provisioner/provision_azure_initializer.go:134]: Add more context to the error.

File: test/provisioner/provision_azure_initializer.go [hunk 4]
```diff
@@ -109,31 +118,41 @@ func initAzureProperties(properties map[string]string) error {
...
 	}
 
-	groupsClient := resources.NewGroupsClient(AzureProps.SubscriptionID)
-	groupsClient.Authorizer = authorizer
-
-	vnetClient := network.NewVirtualNetworksClient(AzureProps.SubscriptionID)
-	vnetClient.Authorizer = authorizer
+	resourcesClientFactory, err := armresources.NewClientFactory(AzureProps.SubscriptionID, cred, nil)
+	if err != nil {
+		return err
```
Comment (surajssd) [inline test/provisioner/provision_azure_initializer.go:128]: Add more context to the error.

File: test/provisioner/provision_azure_initializer.go [hunk 5]
```diff
@@ -109,31 +118,41 @@ func initAzureProperties(properties map[string]string) error {
...
+	virtualMachinesClient := computeClientFactory.NewVirtualMachinesClient()
 
-	AzureProps.ResourceGroupClient = &groupsClient
-	AzureProps.ManagedVnetClient = &vnetClient
-	AzureProps.ManagedSubnetClient = &subnetClient
-	AzureProps.ManagedAksClient = &aksClient
-	AzureProps.ManagedVmClient = &vmClient
+	containerserviceClientFactory, err := armcontainerservice.NewClientFactory(AzureProps.SubscriptionID, cred, nil)
+	if err != nil {
+		log.Fatal(err)
```
Comment (surajssd) [inline test/provisioner/provision_azure_initializer.go:148]: Add more context to the error and return it?

File: test/provisioner/provision_azure.go [hunk 6]
```diff
@@ -211,11 +211,34 @@ func (p *AzureCloudProvisioner) DeleteVPC(ctx context.Context, cfg *envconf.Conf
 func (p *AzureCloudProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Config) error {
 	log.Trace("CreateCluster()")
 
-	pollerResp, err := AzureProps.ManagedAksClient.BeginCreateOrUpdate(
-		context.Background(),
-		AzureProps.ResourceGroupName,
-		AzureProps.ClusterName,
-		armcontainerservice.ManagedCluster{
+	Managedcluster := &armcontainerservice.ManagedCluster{
```
Comment (surajssd) [inline test/provisioner/provision_azure.go:214]: why is this variable capitalized?

File: test/provisioner/provision_azure_initializer.go [hunk 7]
```diff
@@ -86,7 +86,7 @@ func initAzureProperties(properties map[string]string) error {
 	if AzureProps.ClientSecret == "" && !AzureProps.IsAzCliAuth {
 		return errors.New("AZURE_CLIENT_SECRET was not set")
 	}
-	if AzureProps.TenantID == "" && !AzureProps.IsAzCliAuth {
+	if AzureProps.TenantID == "" {
```
Comment (surajssd) [inline test/provisioner/provision_azure_initializer.go:89]: This change should be in the previous commit?

**Review by surajssd — CHANGES_REQUESTED** [review]
I still need to manually try it.

File: test/e2e/azure_test.go [hunk 8]
```diff
@@ -38,22 +37,22 @@ func TestCreatePodWithSecretAzure(t *testing.T) {
...
 
-	for _, vm := range vmList.Values() {
+	for _, vm := range vmList.Value() {
 		if strings.HasPrefix(*vm.Name, prefixName) {
 			// VM found
 			return true
 		}
 	}
-
+	*/
```
[the remark this answers is not in the document]
Comment (kartikjoshi21) [reply test/e2e/azure_test.go]: It was removed in first commit and was again added in third one, as after  changing from compute to armcompute this code was not valid

File: test/provisioner/provision_azure.go (hunk 6 above)
Comment (kartikjoshi21) [reply test/provisioner/provision_azure.go:214]: not required.

File: test/provisioner/provision_azure.go [hunk 9]
```diff
@@ -34,57 +35,60 @@ func init() {
...
-
 		log.Infof("Failed to delete resource group %s:%v.\n", AzureProps.ResourceGroupName, err)
-		return fmt.Errorf("deleting resource group: %w", err)
+		return err
 	}
 
-	log.Infof("Successfully Deleted Resource group %s.\n", AzureProps.ResourceGroupName)
+	_, err = pollerResponse.PollUntilDone(context.Background(), nil)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (kartikjoshi21) [reply test/provisioner/provision_azure.go:70]: its happening in previous step this is just to confirm deletion is done using response.Added context

**Review by bpradipt — APPROVED** [review]
/lgtm
Although I didn't get time to test this.

**Review by surajssd — APPROVED** [review]

**Pull request merged**